src/adapters/philips_hue/api/mod.rs

```diff
@@ -0,0 +1,230 @@
+use std::thread;
+use std::hash::{Hash, SipHasher, Hasher};
+
+
+#[derive(Debug, Hash)]
+struct HueHubApiToken {
+    iv: String,
+    id: String,
+}
+
+impl HueHubApiToken {
+    // This implementation offers 64 bit
+    fn as_hash(&self) -> String {
+        let mut hasher = SipHasher::new();
+        self.hash(&mut hasher);
+        format!("{:016x}", hasher.finish()).to_owned()
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct HueHubApi {
+    pub id: String,
+    pub ip: String,
+    pub token: String,
+}
+
+impl std::fmt::Display for HueHubApi {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        write!(f, "Hue Bridge id:{} at {:?}", self.id, self.ip)
+    }
+}
+
+impl HueHubApi {
+    pub fn new(id: String, ip: String) -> HueHubApi {
+        // TODO: Generate a unique but reproducible access token.
+        // The token must not be guessable by an outsider.
+
+        // TODO: replace with a loooong stored random string
+        let stored_random = String::from("4"); // chosen by fair dice roll.
+        let token_gen = HueHubApiToken { iv: stored_random, id: id.clone() };
+        let token = format!("foxbox-{}", token_gen.as_hash());
+        HueHubApi { id: id, ip: ip, token: token }
+    }
+
+    pub fn get(&self, cmd: String) -> Result<String, Box<Error>> {
+        let url = format!("http://{}/api/{}/{}", self.ip, self.token, cmd);
+        debug!("GET request to Philips Hue bridge {}: {}", self.id, url);
+        let content = http::get(url);
+        debug!("Philips Hue API response: {:?}", content);
+        content
+    }
+
+    pub fn post(&self, cmd: String, data: String) -> Result<String, Box<Error>> {
+        let url = format!("http://{}/api/{}/{}", self.ip, self.token, cmd);
+        debug!("POST request to Philips Hue bridge {}: {} data: {}", self.id, &url, &data);
+        let content = http::post(url, data);
+        debug!("Philips Hue API response: {:?}", content);
+        content
+    }
+
+    pub fn post_unauth(&self, cmd: String, data: String) -> Result<String, Box<Error>> {
+        let url = format!("http://{}/{}", self.ip, cmd);
+        debug!("POST request to Philips Hue bridge {}: {} data: {}", self.id, &url, &data);
+        let content = http::post(url, data);
+        debug!("Philips Hue API response: {:?}", content);
+        content
+    }
+
+    pub fn put(&self, cmd: String, data: String) -> Result<String, Box<Error>> {
+        let url = format!("http://{}/api/{}/{}", self.ip, self.token, cmd);
+        debug!("PUT request to Philips Hue bridge {}: {} data: {}", self.id, &url, &data);
+        let content = http::put(url, data);
+        debug!("Philips Hue API response: {:?}", content);
+        content
+    }
+
+    pub fn is_available(&self) -> bool {
+        let url = format!("http://{}/", self.ip);
+        let content = http::get(url);
+        match content {
+            Ok(value) => {
+                value.contains("hue personal wireless lighting")
```

Comment:
It's probably the least invasive and most expressive way for such a ping. Poking around in the API leaves us just with rather generic responses to work with:

```json
[{"error":{"type":4,"address":"/","description":"method, GET, not available for resource, /"}}]
```

or

```json
[{"error":{"type":1,"address":"/","description":"unauthorized user"}}]
```

neither of which seem expressive enough. If I encode too much of these strings into the check, we in turn need to hope that the API doesn't change its ordering. Perhaps some time I'll come up with something that tries to be intelligent about this.